contracts/IndexStaking.sol

```diff
@@ -0,0 +1,69 @@
+//SPDX-License-Identifier: Unlicense
+pragma solidity ^0.6.6;
+
+import "@openzeppelin/contracts/math/SafeMath.sol";
+import "../libraries/SafeTransfer.sol";
+
+contract IndexStaking {
+    using SafeMath for uint256;
+
+    uint public constant duration = 3 * 365 * 6500; // blocks
+    uint public constant totalSupply = 100000000 * 10**18;
+    uint public constant rewardSupply = 48 * totalSupply / 100;
+
+    address public stakingToken;
+    address public rewardToken;
+
+    uint public staked; // S
+    uint public activeStakeDeposits; // T
+    mapping(address => uint) public stake; // stake
+    mapping(address => uint) public stakedSnapshot; // S0
+
+    uint public accumulatedReward;
+    uint public startBlock;
+
+    constructor(address _stakingTokenAddress, address _rewardTokenAddress) public {
+        stakingToken = _stakingTokenAddress;
+        rewardToken = _rewardTokenAddress;
+        startBlock = block.number;
+    }
+
+    function deposit(uint _amount) public {
+        if (stake[msg.sender] != 0) {
+            withdraw();
+        }
+        SafeTransfer.transferFromERC20(address(stakingToken), msg.sender, address(this), _amount);
+        stake[msg.sender] = _amount;
+        stakedSnapshot[msg.sender] = staked;
+        activeStakeDeposits = activeStakeDeposits.add(_amount);
+    }
+
+    function withdraw() public {
+        uint deposited = stake[msg.sender];
+        uint reward = _calculateReward();
+        if (activeStakeDeposits != 0) {
+            staked = staked.add(reward.div(activeStakeDeposits));
+        }
+        uint userReward = deposited.mul(staked.sub(stakedSnapshot[msg.sender]));
+        activeStakeDeposits = activeStakeDeposits.sub(deposited);
+        stake[msg.sender] = 0;
+        SafeTransfer.sendERC20(address(stakingToken), msg.sender, deposited);
+        SafeTransfer.sendERC20(address(rewardToken), msg.sender, userReward);
+    }
+
+    function withdraw(uint _amount) external {
+        withdraw();
```

Comment:
1. You withdraw everything.
2. Then you deposit amount that you intended to withdraw back...
3. Instead, I think you should deposit the amount left:
```Solidity
allWithdrawn = withdraw();
deposit(allWithdrawn - _amount);
```